feat: enhance `asm` functions to support all* of the Fift-asm syntax

novusnota opened the pull request:
Note, that while it's possible to support the syntax via Ohm without constucting ASTs, it's impossible to support the AST generation without outsourcing that part of the parser to the recursive-descent one, which would also keep track of dictionary of words/instructions and the TVM stack, at least approximating it. [See this for more info](https://github.com/tact-lang/tact/issues/837#issuecomment-2362498612).

P.S.: all* in the title means that we support everything, but with two caveats:
1. Out of place `}` and `{` (when shadowing them, for example) will break the parsing. However, the non-shadowing uses, lists `{ }` and any words containing `}` inside are allowed as long as they are not just a single `}`.
2. Usage of newly created words that use `word` may not parse well, as they might contain the `}` inside of them separated by whitespace, which means that it's another variant of the 1st caveat.

P.P.S: Essentially, our parser only has one flaw, which stems from `{ }` being used as a function block. If we were to use, say, `"""` or `{{{`, the problem would kind of mitigate itself. Just until someone defines the `"""` or `{{{` as a new word in their Fift program, and then we (and FunC), would inevitably fail to parse it. Until we have a [stack-keeping parser](https://github.com/tact-lang/tact/issues/837#issuecomment-2362498612), that is.

<!--
IMPORTANT:
If your PR doesn't close a particular issue, please, create the issue first and describe the whole context: what you're adding/changing and why you're doing so. And only then open the Pull Request, which would close that issue!
-->

## Issue

Closes #857.

## Checklist

- [x] I have updated CHANGELOG.md
- [x] I have added tests to demonstrate the contribution is correctly implemented: this usually includes both positive and negative tests, showing the happy path(s) and featuring intentionally broken cases
- [x] I have run all the tests locally and no test failure was reported
- [x] I have run the linter, formatter and spellchecker
- [x] I did not do unrelated and/or undiscussed refactorings

novusnota committed:
feat: enhance `asm` functions to support all of the Fift-asm syntax

Note, that while it's possible to support the syntax via Ohm without
constucting ASTs, it's impossible to support the AST generation without
outsourcing that part of the parser to the recursive-descent one, which
would also keep track of dictionary of words/instructions and the TVM
stack, at least approximating it. [See this for more
info](https://github.com/tact-lang/tact/issues/837#issuecomment-2362498612).

novusnota committed:
chore: CHANGELOG

novusnota committed:
test: sanity checks

novusnota committed:
test: one more sanity check

The important test with arbitrary words/instructions containing `}` in
them was needed

anton-trunov commented on src/grammar/grammar.ohm:
```diff
@@ -98,8 +98,26 @@ Tact {
+    // Instructions exclude word list begin and end words to ensure correct parse
+    asmInstruction = ~(("[" | "]" | "{" | "}" | "({)" | "(})") ~asmWord) asmWord
+
+    // A chuck of non-whitespace characters forms a word in Fift
```
```suggestion
    // A chunk of non-whitespace characters forms a word in Fift
```

anton-trunov reviewed (changes requested):
Wait, this PR definitely does not close the linked issue as there is no AST model of `asm`-instructions.

novusnota committed:
Update src/grammar/grammar.ohm

Co-authored-by: Anton Trunov <anton@ton.org>

novusnota commented:
Changed the linked issue to #857. I'm willing to return to #837 later, preferably when we start moving to Tact 2.0 and a new parser for it. Alternatively, can tackle it after other remaining issues on 1.6.0 milestone for me to resolve (both here and for docs)

P.S.: On the positive side of #837, our `asm` ASTs will be typechecked right away :)

anton-trunov reviewed (changes requested):
We should definitely add more positive tests from some known Fift-projects.

Can we parse at least some part of `Asm.fif`? 

Some examples can be taken from this list: https://docs.ton.org/develop/smart-contracts/examples#fift-smart-contracts and this list: https://github.com/ton-blockchain/ton/tree/master/crypto/smartcont.

novusnota committed:
test: add Fift libraries from TON Blockchain monorepo

And apply a small fix to the parser — we cannot prohibit shadowing or
removal of instructions because both are used in Fift library and
beyond.

However, that means that out of place `{` and `}` are possible when
shadowing them. See the comment in `embed-fift-fif.tact` about such
case, and why it's ok to move forward with it

novusnota committed:
chore: update snapshots

novusnota commented:
@anton-trunov added Fift library to the test suite and made sure that we don't create unnecessary snapshots for its contents

anton-trunov reviewed (approved):
Amazing job!
